external-crates/move/crates/move-analyzer/src/symbols.rs

```diff
@@ -841,369 +886,732 @@ impl Symbols {
-                    eprintln!("typed AST compilation failed");
-                    return Ok((files, vec![]));
-                }
-            };
-            eprintln!("compiled to typed AST");
-            let (compiler, typed_program) = compiler.into_ast();
-            typed_ast = Some(typed_program.clone());
-            eprintln!("compiling to bytecode");
-            let compilation_result = compiler.at_typing(typed_program).build();
-            let (units, diags) = match compilation_result {
-                Ok(v) => v,
-                Err(diags) => {
-                    let failure = false;
-                    diagnostics = Some((diags, failure));
-                    eprintln!("bytecode compilation failed");
-                    return Ok((files, vec![]));
-                }
-            };
-            // warning diagnostics (if any) since compilation succeeded
-            if !diags.is_empty() {
-                // assign only if non-empty, otherwise return None to reset previous diagnostics
-                let failure = false;
+    let build_plan = BuildPlan::create(resolution_graph)?;
+    let mut parsed_ast = None;
+    let mut typed_ast = None;
+    let mut diagnostics = None;
+    build_plan.compile_with_driver(&mut std::io::sink(), |compiler| {
+        // extract expansion AST
+        let (files, compilation_result) = compiler.run::<PASS_PARSER>()?;
+        let (_, compiler) = match compilation_result {
+            Ok(v) => v,
+            Err(diags) => {
+                let failure = true;
                 diagnostics = Some((diags, failure));
+                eprintln!("parsed AST compilation failed");
+                return Ok((files, vec![]));
             }
-            eprintln!("compiled to bytecode");
-            Ok((files, units))
-        })?;
-
-        let mut ide_diagnostics = lsp_empty_diagnostics(&file_name_mapping);
-        if let Some((compiler_diagnostics, failure)) = diagnostics {
-            let lsp_diagnostics = lsp_diagnostics(
-                &compiler_diagnostics.into_codespan_format(),
-                &files,
-                &file_id_mapping,
-                &file_name_mapping,
-            );
-            // start with empty diagnostics for all files and replace them with actual diagnostics
-            // only for files that have failures/warnings so that diagnostics for all other files
-            // (that no longer have failures/warnings) are reset
-            ide_diagnostics.extend(lsp_diagnostics);
-            if failure {
-                // just return diagnostics as we don't have typed AST that we can use to compute
-                // symbolication information
-                debug_assert!(typed_ast.is_none());
-                return Ok((None, ide_diagnostics));
+        };
+        eprintln!("compiled to parsed AST");
+        let (compiler, parsed_program) = compiler.into_ast();
+        parsed_ast = Some(parsed_program.clone());
```

Comment:
Can `parsed_ast` evaluate to `None` in a way that matters? It seems like if we succeed, `parsed_ast` is a value, and if not we return. So maybe this can just be a value without option and we assign it here?

--see related comment below

Reply:
> Can parsed_ast evaluate to None in a way that matters?

This was written in my early days of rust, but even now I am not sure what the best pattern is to return a value from a lambda to the function enclosing this lambda... Suggestions welcome!